### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by mcalus3**

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module.exports = {
+  extends: [require.resolve('@backstage/cli/config/eslint')],
```
Comment (punkle) [inline .eslintrc.js:2]: I would be inclined excluding this dependency if at all possible. I dont think it is worth the weight if just for eslint config.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    "start": "backstage-cli plugin:serve",
+    "lint": "backstage-cli lint",
+    "test": "backstage-cli test",
+    "diff": "backstage-cli plugin:diff",
+    "prepack": "backstage-cli prepack",
+    "postpack": "backstage-cli postpack",
+    "clean": "backstage-cli clean"
+  },
+  "dependencies": {
+    "@backstage/catalog-model": "^0.2.0",
```
Comment (lewtakm) [inline package.json:24]: I would suggest to bump plugins version to the latest ( catalog, cli etc.)

File: src/AuthPopup.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    } else {
+      const currentUrl = new URL(window.location.href);
+      var redirectUrl = currentUrl.searchParams.get('url');
+      const currentUrlWithoutParams = `${window.location.origin}${window.location.pathname}`;
+      window.location.replace(
+        `${redirectUrl}?url=${currentUrlWithoutParams}&vouch-failcount=&X-Vouch-Token=&error=&rd=${currentUrlWithoutParams}`,
+      );
+    }
+  }, []);
+  return <div></div>;
```
Comment (lewtakm) [inline src/AuthPopup.tsx:31]: Can't we just return null there?

File: src/fetchWithVouchRedirect.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        console.log(event.data, event.origin, window.location.origin);
+        if (event.data === 'closePopup') {
+          windowObjectReference.close();
+        }
+      },
+      false,
+    );
+    await new Promise(res => {
+      var timer = setInterval(function () {
```
Comment (lewtakm) [inline src/fetchWithVouchRedirect.ts:50]: Please use let instead of var

File: README.md [hunk 5]
```diff
@@ -1 +1 @@
-# fetch-with-redirected-login
+# Vouch redirect Plugin for Backstage
```
Comment (punkle) [inline README.md:1]: Is this definitely specific for vouch?

Can we fill this out a little?

e.g.

Explain what it does.
Explain that the following settings are required in vouch and how to set them:
```
  vouch:
    cookie:
      sameSite: none
    domains: ["hostofsite", ...domains]
```
Explain what the javascript interface is and how to include it in a project. e.g.
```
import { fetchVouchProtectedResource } from 'fetch-with-redirected-login'
fetchVouchProtectedResource('https://protected-service/hello/world')
```

File: src/fetchWithVouchRedirect.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    } else {
+      windowObjectReference.focus();
+    }
+    window.addEventListener(
+      'message',
+      event => {
+        if (event.origin !== window.location.origin) {
+          return;
+        }
+        console.log(event.data, event.origin, window.location.origin);
```
Comment (lewtakm) [inline src/fetchWithVouchRedirect.ts:42]: Console.log :)

File: src/fetchWithVouchRedirect.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    return await fetch(initialUrl, {
+      credentials: 'include',
+    });
+  }
+
+  async function obtainAuthCookie(initialUrl: string, authUrl?: string) {
+    const popupUrl = authUrl
+      ? `${window.location.origin}/authpopup?url=${authUrl}`
+      : initialUrl;
+    if (windowObjectReference == null || windowObjectReference.closed) {
```
Comment (lewtakm) [inline src/fetchWithVouchRedirect.ts:22]: Shouldn't it be === instead of == ?

File: src/fetchWithVouchRedirect.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      windowObjectReference = window.open(
+        popupUrl,
+        'PromoteFirefoxWindowName',
+        'resizable,scrollbars,status',
+      );
+      if (windowObjectReference == null || windowObjectReference.closed) {
```
Comment (lewtakm) [inline src/fetchWithVouchRedirect.ts:28]: as above

File: src/setupTests.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+import '@testing-library/jest-dom';
```
Comment (punkle) [inline src/setupTests.ts:1]: Is it possible to add a basic github workflow with linting and possibly some low hanging fruit testing?

File: src/fetchWithVouchRedirect.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
+var windowObjectReference: Window | null = null;
```
Comment (lewtakm) [inline src/fetchWithVouchRedirect.ts:1]: Use let instead of var

File: src/index.ts [hunk 11]
```diff
@@ -0,0 +1,2 @@
+export { fetchWithVouchRedirect } from './fetchWithVouchRedirect';
+export { AuthPopup } from './AuthPopup';
```
Comment (punkle) [inline src/index.ts:2]: does this need to be exported?

Comment (mcalus3) [reply src/index.ts:2]: AuthPopup has to be exported to be able to be mounted in the backstage App router

Comment (punkle) [reply src/index.ts:2]: ah ok got it.

File: README.md (hunk 5 above)
Comment (mcalus3) [reply README.md:1]: I had to add vouch-specific things there to enable auto-closing feature (redirecting popup through our own site first)

File: .eslintrc.js (hunk 1 above)
Comment (mcalus3) [reply .eslintrc.js:2]: All unneeded dependencies are to be excluded

File: package.json (hunk 2 above)
Comment (mcalus3) [reply package.json:24]: All unneeded dependencies are yet to be excluded, rest of them has to be updated to the latest version

Comment (mcalus3) [reply package.json]: Removed all unneeded dependencies, bumped cli dep

File: src/AuthPopup.tsx (hunk 3 above)
Comment (mcalus3) [reply src/AuthPopup.tsx]: done

File: src/fetchWithVouchRedirect.ts (hunk 4 above)
Comment (mcalus3) [reply src/fetchWithVouchRedirect.ts]: done

File: src/fetchWithVouchRedirect.ts (hunk 6 above)
Comment (mcalus3) [reply src/fetchWithVouchRedirect.ts]: done

File: src/fetchWithVouchRedirect.ts (hunk 7 above)
Comment (mcalus3) [reply src/fetchWithVouchRedirect.ts]: done

File: src/fetchWithVouchRedirect.ts (hunk 8 above)
Comment (mcalus3) [reply src/fetchWithVouchRedirect.ts]: done

File: src/fetchWithVouchRedirect.ts (hunk 10 above)
Comment (mcalus3) [reply src/fetchWithVouchRedirect.ts]: done

File: src/setupTests.ts (hunk 9 above)
Comment (mcalus3) [reply src/setupTests.ts:1]: About testing - this code i 99% about opening popups, sending messages between windows and opening external sites, which all aren't supported by testing frameworks. Only thing that don't have to be mocked is string concatenation (for url creation), in which tests doesn't give any benefits.

File: README.md (hunk 5 above)
Comment (mcalus3) [reply README.md:1]: Added a readme file

File: src/setupTests.ts (hunk 9 above)
Comment (punkle) [reply src/setupTests.ts:1]: At its most basic I thought we could add something like this:

When the fetch does not fail, it asserts that the response object is returned.
When the fetch does fail it asserts that obtainAuthCookie is called.
If obtainAuthCookie succeeds, it asserts that fetch is called a second time and returns the response object returned by the second fetch
If obtainAuthCookie fails, it asserts the caller is surfaced a reasonable error.
If the second fetch fails, it asserts the caller is surfaced a reasonable error.

I think you are right about testing obtainAuthCookie. I am not sure how that function could be tested.

**Pull request merged**